nuxeo-features/nuxeo-annotation/nuxeo-annotation-core/src/main/java/org/nuxeo/ecm/annotation/AnnotationServiceImpl.java

```diff
@@ -56,11 +59,40 @@
 
     protected static final String GET_ANNOTATIONS_FOR_DOC_PAGEPROVIDER_NAME = "GET_ANNOTATIONS_FOR_DOCUMENT";
 
+    protected static final String ANNOTATIONS_PLACELESS_STORAGE_PROPERTY = "nuxeo.annotations.placeless.storage";
+
+    protected static final String HIDDEN_FOLDER_TYPE = "HiddenFolder";
+
     @Override
     public Annotation createAnnotation(CoreSession session, Annotation annotation) {
-        // Create annotation as a placeless document
+
+        ConfigurationService configurationService = Framework.getService(ConfigurationService.class);
+        boolean annotationPlaceless = configurationService.isBooleanPropertyTrue(
+                ANNOTATIONS_PLACELESS_STORAGE_PROPERTY);
+
         return CoreInstance.doPrivileged(session, s -> {
-            DocumentModel annotationModel = s.createDocumentModel(null, ANNOTATION_NAME, ANNOTATION_DOC_TYPE);
+            String path = null;
+            if (!annotationPlaceless) {
+                // Create or retrieve the folder to store the annotation.
+                // If the document is under a domain, the folder is a child of this domain.
+                // Otherwise, it is a child of the root document.
+                DocumentModel annotatedDoc = s.getDocument(new IdRef(annotation.getDocumentId()));
+                String parentPath = s.getRootDocument().getPathAsString();
+                if (annotatedDoc.getPath().segmentCount() > 1) {
+                    parentPath += annotatedDoc.getPath().segment(0);
+                }
+                // Create a folder for each version of the annotated document
```

Comment:
Let's not do that, it's not needed and in theory it's prone to race conditions between two nodes.